src/fidesops/ops/service/strategy.py

```diff
@@ -0,0 +1,77 @@
+from __future__ import annotations
+
+import logging
+from abc import ABC
+from typing import Any, Dict, Generic, List, Optional, Type, TypeVar
+
+from pydantic import ValidationError
+
+from fidesops.ops.common_exceptions import NoSuchStrategyException
+from fidesops.ops.common_exceptions import ValidationError as FidesopsValidationError
+from fidesops.ops.schemas.saas.strategy_configuration import StrategyConfiguration
+
+logger = logging.getLogger(__name__)
+T = TypeVar("T", bound="Strategy")
+C = TypeVar("C", bound=StrategyConfiguration)
+
+
+def _find_strategy_subclass(
+    cls: Type[Strategy], strategy_name: str
+) -> Optional[Type[Strategy]]:
+    if hasattr(cls, "name") and cls.name == strategy_name:
+        return cls
+    for sub in cls.__subclasses__():
+        found = _find_strategy_subclass(sub, strategy_name)
+        if found is not None:
+            return found
+    return None
+
+
+def _find_all_strategy_subclasses(
+    cls: Type[T], subs: List[Type[T]] = None
+) -> List[Type[T]]:
+    if not subs:
+        subs = []
+    if hasattr(cls, "name"):
+        subs.append(cls)
+    for sub in cls.__subclasses__():
+        _find_all_strategy_subclasses(sub, subs)
+    return subs
+
+
+class Strategy(ABC, Generic[C]):
+    """Abstract base class for strategies"""
+
+    name: str
+    configuration_model: Type[C]
+
+    @classmethod
+    def get_strategy(
+        cls: Type[T],
+        strategy_name: str,
+        configuration: Dict[str, Any],
+    ) -> T:
+        """
+        Returns the strategy given the name and configuration.
+        Raises NoSuchStrategyException if the strategy does not exist
+        """
+
+        strategy_class = _find_strategy_subclass(cls, strategy_name)
+
+        if strategy_class is None:
+            valid_strategies = ", ".join(
+                [sub.name for sub in _find_all_strategy_subclasses(cls)]
+            )
```

Comment:
Do you have an example of when `get_strategy` would not find the strategy class, and we'd have to resort to calling `_find_all_strategy_subclasses`?   Since `_find_all_strategy_subclasses` returns an empty list on your latest commit, I don't think we have tests that cover this use case.

Reply:
good point - `test_retrieve_nonexistent_strategy` does hit that code path, but it wasn't testing for the string populated by `_find_all_strategy_subclasses`, so i'll update it to verify that string is being populated correctly.

Reply:
ah thanks!